File: 367-valid-perfect-square/367-valid-perfect-square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        l = 1
        r = 2 ** 31 -1
        while l < r:
            mid = (l + r) / 2
            ans = mid * mid
            if ans == num:
                return True
            if ans > num:
                r = mid - 1
            else:
                l = mid + 1
        if l * l == num:
            return True
        return False

File: 367-valid-perfect-square/367-valid-perfect-square.cpp
class Solution {
public:
    bool isPerfectSquare(int num) {
        long l = 1;
        long r = (num / 2) + 1;
        
        while(l <= r) {
            long mid = (l + r) / 2;
            long long ans = mid * mid;
            if(ans == num) return true;
            if(ans > num) r = mid - 1;
            else l = mid + 1;
        }

        return false;
    }
};